File: 139.js
// 139. Word Break
// https://leetcode.com/problems/word-break/

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */

// Divide-and-conquer
var wordBreak = function (s, wordDict) {};

// Dynamic Programming
// time complexity: O(w + n^2 * w) = O(n^2 * w) || space complexity: O(n)
var wordBreak = function (s, wordDict) {
  const wordSet = new Set(wordDict);

  const length = s.length;

  const memo = new Array(length + 1).fill(false);
  memo[0] = true;

  let target;
  for (let i = 0; i < length; i++) {
    for (let j = i; j < length; j++) {
      target = s.slice(i, j + 1);

      if (wordSet.has(target) && memo[i]) {
        memo[j + 1] = memo[i];
      }
    }
  }

  return memo[length];
};

File: LeetCode - Top Interview 150/139.py
# 139. Word Break
# https://leetcode.com/problems/word-break/


from typing import List


class Solution:
    # time complexity: O(nmk) (n: length of s, m: length of wordDict, k: average length of each word in wordDict)
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        length = len(s)

        memo = [False] * (length + 1)
        memo[0] = True

        for i in range(1, length + 1):
            for word in wordDict:
                prev_index = i - len(word)

                if prev_index < 0:
                    continue

                if not memo[prev_index]:
                    continue

                if s[prev_index:i] == word:
                    memo[i] = memo[prev_index]

        return memo[-1]
